calico_kubernetes/calico_kubernetes.py

```diff
@@ -97,6 +147,223 @@ def _configure_interface(self):
+
+        # Create the veth, move into the container namespace, add the IP and
+        # set up the default routes.
+        # Update Endpoint in felix
+        pid = info["State"]["Pid"]
+        ep = self._endpoint_complete(ep, pid, ip, interface)
+
+        # Let the caller know what endpoint was created.
+        return ep
+
+    def _container_remove(self, container_name):
+        """
+        Remove the indicated container on this host from Calico networking
+
+        :param container_name: THe name or partial id of the container
+        """
+        # Resolve the name to ID.
+        workload_id = self._get_container_id(container_name)
+
+        # Find the endpoint ID. We need this to find any ACL rules
+        try:
+            endpoint = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=ORCHESTRATOR_ID,
+                workload_id=workload_id
+            )
+        except KeyError:
+            print("Container %s doesn't contain any endpoints" % container_name)
+            sys.exit(1)
+
+        # Remove any IP address assignments that this endpoint has
+        for net in endpoint.ipv4_nets | endpoint.ipv6_nets:
+            assert(net.size == 1)
+            ip = net.ip
+            self._unassign_address_in_pool_IPAM(ip)
+
+        # Remove the endpoint
+        netns.remove_veth(endpoint.name)
+
+        # Remove the container from the datastore.
+        self._datastore_client.remove_workload(hostname, ORCHESTRATOR_ID, workload_id)
+
+        print("Removed Calico interface from %s" % container_name)
+
+    def _validate_container_add(self, info):
+        container_id = info["Id"]
+
+        # Check if the container already exists
+        try:
+            _ = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=ORCHESTRATOR_ID,
+                workload_id=container_id
+            )
+        except KeyError:
+            # Calico doesn't know about this container.  Continue.
+            pass
+        else:
+            # Calico already set up networking for this container.  Since we got
+            # called with an IP address, we shouldn't just silently exit, since
+            # that would confuse the user: the container would not be reachable on
+            # that IP address.
+            print("This container has already been configured with Calico Networking.")
+            sys.exit(1)
+
+        # Check the container is actually running.
+        if not info["State"]["Running"]:
+            print("The container is not currently running.")
+            sys.exit(1)
+
+        # We can't set up Calico if the container shares the host namespace.
+        if info["HostConfig"]["NetworkMode"] == "host":
+            print("Can't add the container to Calico because it is running NetworkMode = host.")
+            sys.exit(1)
+
+    def _assign_address_in_pool_IPAM(self, ip):
```

Comment:
Just a note on names - I think `_assign_address(...)` and `_free_address(...)` are more suitable.